dev/NavigationView/NavigationView.cpp

```diff
@@ -379,10 +427,354 @@ void NavigationView::OnApplyTemplate()
     UpdateBackAndCloseButtonsVisibility();
     UpdateSingleSelectionFollowsFocusTemplateSetting();
     UpdateNavigationViewUseSystemVisual();
-    PropagateNavigationViewAsParent();
     UpdatePaneVisibility();
     UpdateVisualState();
     UpdatePaneTitleMargins();
+    SyncItemTemplates();
+}
+
+void NavigationView::UpdateRepeaterItemsSource(bool forceSelectionModelUpdate)
+{
+    auto dataSource = MenuItemsSource();
+    if (!dataSource)
+    {
+        dataSource = MenuItems();
+        UpdateSelectionForMenuItems();
+    }
+
+    // Selection Model has same representation of data regardless
+    // of pane mode, so only update if the ItemsSource data itself
+    // has changed.
+    if (forceSelectionModelUpdate){
+        m_selectionModel.Source(dataSource);
+    }
+
+    if (IsTopNavigationView())
+    {
+        UpdateLeftRepeaterItemSource(nullptr);
+        UpdateTopNavRepeatersItemSource(dataSource);
+    }
+    else
+    {
+        UpdateTopNavRepeatersItemSource(nullptr);
+        UpdateLeftRepeaterItemSource(dataSource);
+    }
+
+    if (IsTopNavigationView())
+    {
+        InvalidateTopNavPrimaryLayout();
+        UpdateSelectedItem();
+    }
+}
+
+void NavigationView::UpdateLeftRepeaterItemSource(const winrt::IInspectable& items)
+{
+    UpdateItemsRepeaterItemsSource(m_leftNavRepeater.get(), items);
+}
+
+void NavigationView::UpdateTopNavRepeatersItemSource(const winrt::IInspectable& items)
+{
+    // Unhook TopNav repeater
+    UpdateItemsRepeaterItemsSource(m_topNavRepeater.get(), nullptr);
+    UpdateItemsRepeaterItemsSource(m_topNavRepeaterOverflowView.get(), nullptr);
+
+    // Change data source and setup vectors
+    m_topDataProvider.SetDataSource(items);
+
+    // rebinding
+    if (items)
+    {
+        UpdateItemsRepeaterItemsSource(m_topNavRepeater.get(), m_topDataProvider.GetPrimaryItems());
+        UpdateItemsRepeaterItemsSource(m_topNavRepeaterOverflowView.get(), m_topDataProvider.GetOverflowItems());
+    }
+    else
+    {
+        UpdateItemsRepeaterItemsSource(m_topNavRepeater.get(), nullptr);
+        UpdateItemsRepeaterItemsSource(m_topNavRepeaterOverflowView.get(), nullptr);
+    }
+}
+
+void NavigationView::UpdateItemsRepeaterItemsSource(const winrt::ItemsRepeater& ir,
+    const winrt::IInspectable& itemsSource)
+{
+    if (ir)
+    {
+        auto oldItemsSource = ir.ItemsSource();
+        if (oldItemsSource != itemsSource)
+        {
+            ir.ItemsSource(itemsSource);
+        }
+    }
+}
+
+void NavigationView::OnNavigationViewItemIsSelectedPropertyChanged(const winrt::DependencyObject& sender, const winrt::DependencyProperty& args)
+{
+    auto nvib = sender.try_as<winrt::NavigationViewItemBase>();
+    if (nvib)
+    {
+        // Check whether the container that triggered this call back is the selected container
+        bool newItemIsSelectedItem = IsContainerTheSelectedItemInTheSelectionModel(nvib);
```

Comment:
The naming makes it a little hard to understand. Perhaps name them as isIndexSelectedInModel and isIndexSelectedInContainer or something along those lines ?